Fix size 65536 HybiFrames

divinity76 opened the pull request:
off-by-1: previously, hybiframes with a 65536 bytes payload would generate a 16 bit size header signaling 0 bytes (pack("n", 65536) => "\x00\x00"), instead of using a 64bit size header.

nit: instead of is_int && in_array, we can use in_array's strict mode.

nit: $this->offset_mask and $this->offset_payload were initialized-to-null in constructor and encode, then written internally in getPayloadOffset,
 then written in encode() again.

nit: use pack('J') instead of manually constructing big_endian_u64 (undocumented funfact, J is not available in 32bit php: https://github.com/php/doc-en/pull/4644 , that may be why the original code did not use J)

nit: getInitialLength() was calculating the result twice internally.

<strike>nit: composer.json php version had not been updated since 8.0.2 </strike> wrong, my bad, reverted.

<strike>nit: use random_bytes instead of openssl_random_pseudo_bytes(), the original code was written for php5 where random_bytes did not exist.</strike> moved to dedicated PR: https://github.com/chrome-php/wrench/pull/22

divinity76 committed:
fix size 65536 HybiFrame + nitpicks

off-by-1: previously, hybiframes with a 65536 bytes payload
would generate a 16 bit size header signaling 0 bytes (pack("n", 65536) => "\x00\x00"),
instead of using a 64bit size header.

nit: composer.json php version had not been updated since 8.0.2

nit: instead of is_int && in_array, we can use in_array's strict mode.

nit: $this->offset_mask and $this->offset_payload were
initialized-to-null in constructor and encode, then written internally in getPayloadOffset,
 then written in encode() again.

nit: use pack('J') instead of manually constructing big_endian_u64 (undocumented funfact, J is not available in 32bit php: https://github.com/php/doc-en/pull/4644 ,
that may be why the original code did not use J)

nit: use random_bytes instead of openssl_random_pseudo_bytes(),
the original code was written for php5 where random_bytes did not exist.

nit: getInitialLength() was calculating the result twice internally.

divinity76 committed:
StyleCI

divinity76 commented on src/Frame/HybiFrame.php:
```diff
@@ -69,58 +69,60 @@ class HybiFrame extends Frame
+                | self::BITFIELD_FINAL
         );
 
-        $masked_bit = (self::BITFIELD_MASKED & ($this->masked ? \PHP_INT_MAX : 0));
```
this was just unnecessarily complex / difficult to read.

divinity76 commented on src/Frame/HybiFrame.php:
```diff
@@ -69,58 +69,60 @@ class HybiFrame extends Frame
+        // FIN + opcode byte
         $this->buffer[self::BYTE_HEADER] = \chr(
             (self::BITFIELD_TYPE & $this->type)
-            | (self::BITFIELD_FINAL & \PHP_INT_MAX)
```
(non-negative-integer & \PHP_INT_MAX)
is a no-op 🤔 not sure what the original intention was.

GrahamCampbell commented on composer.json:
```diff
@@ -11,7 +11,7 @@
     ],
     "require": {
-        "php": "^7.4.15 || ^8.0.2",
+        "php": "^7.4.15 || ^8.0.2 || ^8.1.0 || ^8.2.0 || ^8.3.0 || ^8.4.0",
```
This change is incorrect. ^8.0.2 already allows up to 8.99999.99999

GrahamCampbell commented on src/Frame/HybiFrame.php:
```diff
@@ -131,13 +133,7 @@ public function encode(string $payload, int $type = Protocol::TYPE_TEXT, bool $m
-            // So we pack it into 32 bits
-            return \pack('N', \sha1(\spl_object_hash($this).\mt_rand(0, \PHP_INT_MAX).\uniqid('', true), true));
-        }
+        return \random_bytes(4);
```
Is this change strictly necessary for the fix? If not, I'd rather it be moved to a separate atomic PR.

divinity76 replied:
moved to https://github.com/chrome-php/wrench/pull/22

divinity76 committed:
PR feedback

https://github.com/chrome-php/wrench/pull/24/files#r2070000317
and
https://github.com/chrome-php/wrench/pull/24/files#r2069999223
